### PR TITLE
GEMM with Multiple Source, GEMM+Bias+Add+FastGeLU example and ckProfiler

### PR DESCRIPTION
**Pull request by asroy**
* add GEMM with multi-source kernel
* use above item to implement  GEMM+Add+Add+Fast_GeLU device operation: example and ckProfiler
* some changes to utility code

### CODE REVIEW

File: example/CMakeLists.txt [hunk 1]
```diff
@@ -33,7 +33,7 @@ function(add_example_executable_no_testing EXAMPLE_NAME FILE_NAME)
     add_executable(${EXAMPLE_NAME} ${FILE_NAME})
     target_link_libraries(${EXAMPLE_NAME} PRIVATE host_tensor)
     add_dependencies(examples ${EXAMPLE_NAME})
-endfunction(add_example_executable EXAMPLE_NAME)
+endfunction(add_example_executable_no_testing EXAMPLE_NAME)
```
Comment (rosenrodt) [inline example/CMakeLists.txt:36]: Nice catch

**Review by rosenrodt — COMMENTED** [review]
Please resolve formatting issue. Otherwise looks good

File: include/ck/tensor_operation/gpu/block/thread_group_tensor_slice_transfer_v7.hpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+#include "tensor_descriptor.hpp"
+#include "tensor_descriptor_helper.hpp"
+#include "cluster_descriptor.hpp"
+#include "threadwise_tensor_slice_transfer_v7.hpp"
+
+namespace ck {
+
+// this version does following things to avoid scratch memory issue
+// 1. Use StaticallyIndexedArray instead of C array for thread buffer
+// 2. ThreadwiseTensorSliceTransfer_v3 does not keep reference to tensor descriptor
```
Comment (rosenrodt) [inline include/ck/tensor_operation/gpu/block/thread_group_tensor_slice_transfer_v7.hpp:13]: Could you update comments to reflect what's different in `v7`?

File: include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+          typename EGridDescriptor_MBlock_MPerBlock_NBlock_NPerBlock,
+          typename Block2ETileMap,
+          bool HasMainKBlockLoop>
+__global__ void
+#if CK_USE_LAUNCH_BOUNDS
+    __launch_bounds__(CK_MAX_THREAD_PER_BLOCK, CK_MIN_BLOCK_PER_CU)
+#endif
+        kernel_gemm_multiple_d_xdl_cshuffle(const FloatAB* __restrict__ p_a_grid,
+                                            const FloatAB* __restrict__ p_b_grid,
+                                            FloatDsPointer p_ds_grid,
```
Comment (rosenrodt) [inline include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp:37]: @rocking5566 and I were discussing whether losing `__restrict__` qualifier would require the compiler to apply strict aliasing rule which would hurt performance a bit. We don't have an answer yet but something to take note of nevertheless.

File: include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp [hunk 4]
```diff
@@ -0,0 +1,750 @@
...
+                                              arg.ds_grid_desc_mblock_mperblock_nblock_nperblock_,
+                                              arg.e_grid_desc_mblock_mperblock_nblock_nperblock_,
+                                              arg.block_2_etile_map_);
+            };
+
+            float ave_time = 0;
+
+            if(GridwiseGemm::CalculateHasMainKBlockLoop(K))
+            {
+                ave_time = launch_kernel(integral_constant<bool, true>{});
```
Comment (rosenrodt) [inline include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp:618]: Neat

**Review by rosenrodt — DISMISSED** [review]

File: include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp (hunk 3 above)
Comment (asroy) [reply include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp:37]: you mean if we remove ```__restrict__```, compiler may not do certain optimization, right?

Just curious, is there any benefit in removing this keyword?

File: include/ck/tensor_operation/gpu/block/thread_group_tensor_slice_transfer_v7.hpp (hunk 2 above)
Comment (asroy) [reply include/ck/tensor_operation/gpu/block/thread_group_tensor_slice_transfer_v7.hpp:13]: Done

File: include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp (hunk 3 above)
Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp:37]: Removing `__restrict__` means 2 pointers of some type `T` (where the strict aliasing rule applies) could alias each other so kernel would re-read the data from one `T` array after writing to another `T` array.

We can borrow idea from Pytorch, which means to cast pointers to `T* __restrict__` upon `make_dynamic_buffer()`

https://github.com/pytorch/pytorch/blob/73f6601cfcd9b869dde5613ab9c8aec2da28527b/aten/src/ATen/core/TensorAccessor.h#L28-L34

Comment (asroy) [reply include/ck/tensor_operation/gpu/device/device_gemm_multiple_d_xdl_cshuffle.hpp:37]: >> We can borrow idea from Pytorch, which means to cast pointers to T* __restrict__ upon make_dynamic_buffer()

What is the benefit of doing that?

**Review by zjing14 — APPROVED** [review]

**Pull request merged**